File: frontend/src/lib/maiaEncoding.ts
/**
 * maiaEncoding.ts — original MIT glue for the Maia-3 ("Chessformer") ONNX model:
 * board->tensor encoding, legal-move policy masking + softmax, expected-score, and
 * the ELO-ladder input mechanism (MAIA-03).
 *
 * This module is NOT derived from CSSLab's AGPL-licensed reference client source
 * (`maia-platform-frontend`'s `MaiaEngineContext`/encoding utilities). It is written
 * from scratch against the CONFIRMED tensor contract recorded in
 * `.planning/phases/151-maia-in-the-browser-all-position-surfaces/151-MAIA-CONTRACT.md`,
 * which was produced by loading the real `maia3_simplified.onnx` and inspecting its
 * declared I/O + running real inference (Plan 151-01). See that file for the full
 * evidence trail; the load-bearing facts are summarized in the comments below.
 */

import { Chess } from 'chess.js';

// ─── Board encoding constants (CONTRACT §a) ────────────────────────────────────

/** 8x8 board, one token per square. */
export const NUM_SQUARES = 64;

/** 12 one-hot piece-occupancy planes per square (no history — n=0, "simplified" export). */
export const PLANES_PER_SQUARE = 12;

/**
 * Confirmed 12-plane order (CONTRACT §a): white P,N,B,R,Q,K (indices 0-5),
 * black p,n,b,r,q,k (indices 6-11).
 */
const PIECE_PLANE_ORDER = ['P', 'N', 'B', 'R', 'Q', 'K', 'p', 'n', 'b', 'r', 'q', 'k'] as const;

// ─── ELO ladder (CONTRACT §c; range widened per UAT quick 260705-bm3) ────────────

/**
 * Lower bound of the ELO ladder. maiachess.com presents the full 600–2600 human
 * range, so we match it (UAT quick 260705-bm3). NOTE: Plan 151-01 behaviorally
 * validated the model only across 1100–2000 (monotonically rising draw-rate sweep);
 * 600–1000 are extrapolation BELOW that validated band — kept because Maia-3 takes a
 * continuous ELO input and maiachess.com itself displays this range.
 */
const MAIA_ELO_LADDER_MIN = 600;

/**
 * Upper bound of the ELO ladder. See MIN: 2100–2600 are extrapolation ABOVE the
 * 151-01-validated 1100–2000 band, matching maiachess.com's displayed range.
 */
const MAIA_ELO_LADDER_MAX = 2600;

/** Step size between rungs — named per CLAUDE.md "no magic numbers". */
const MAIA_ELO_LADDER_STEP = 100;

/**
 * The ELO ladder: 600..2600 in steps of 100 (21 rungs), matching maiachess.com's
 * presented range (UAT quick 260705-bm3). The behaviorally-validated sub-band is
 * 1100..2000 (151-MAIA-CONTRACT.md §c / Plan 151-01); the ends are extrapolation.
 */
export const MAIA_ELO_LADDER: readonly number[] = Array.from(
  { length: (MAIA_ELO_LADDER_MAX - MAIA_ELO_LADDER_MIN) / MAIA_ELO_LADDER_STEP + 1 },
  (_, i) => MAIA_ELO_LADDER_MIN + i * MAIA_ELO_LADDER_STEP,
);

// ─── Policy vocabulary (CONTRACT §d) ────────────────────────────────────────────

/** Total flat policy-logit vocabulary size, confirmed by real inference (CONTRACT §d). */
export const POLICY_VOCAB_SIZE = 4352;

/** Base (non-underpromotion) move-index space: every (from, to) square pair. */
const BASE_VOCAB_SIZE = NUM_SQUARES * NUM_SQUARES; // 4096

/**
 * Underpromotion piece lanes reserved per destination square. 4352 - 4096 = 256 =
 * 64 destination squares x 4 piece lanes — this is the only integer decomposition
 * of the confirmed vocab size that cleanly separates "every from->to pair" (queen
 * promotions collapse into the base lane, since a pawn reaching the back rank always
 * promotes and 'to' alone disambiguates it from a non-promoting move) from "explicit
 * underpromotion, keyed by destination + piece only" (origin is unambiguous from
 * board state: exactly one pawn can reach a given back-rank square in one move).
 *
 * ASSUMPTION (flagged, not silently guessed): the CONTRACT confirms the vocab SIZE
 * (4352) and the general key scheme ("from+to+promotion"), but not CSSLab's literal
 * index ORDER for `all_moves_maia3.json` — reconstructing their exact enumeration
 * order without copying their AGPL data file is the phase's acknowledged open risk
 * (151-MAIA-CONTRACT.md §d: "Wave 2 must ship our own vocabulary/index table").
 * This scheme is deterministic and internally consistent (round-trips correctly
 * inside this module), but its alignment with the REAL model's index order is
 * unverified until VALID-01's real-ONNX cross-check (Plan 151-06) — see
 * 151-04-SUMMARY.md "Known Limitations" for the explicit follow-up.
 */
const UNDERPROMOTION_PIECE_LANES = ['q', 'r', 'b', 'n'] as const;
type PromotionPiece = (typeof UNDERPROMOTION_PIECE_LANES)[number];

// ─── Types ──────────────────────────────────────────────────────────────────────

/** WDL probability vector (post-softmax), in the natural W/D/L reading order. */
export interface WdlVector {
  win: number;
  draw: number;
  loss: number;
}

// ─── Square indexing (CONTRACT §a) ──────────────────────────────────────────────

/**
 * Maps an algebraic square (e.g. "e4") to the CONTRACT's confirmed token index:
 * `s = row*8 + file`, `row = rank - 1` (a1 = 0, h8 = 63).
 */
export function squareIndex(square: string): number {
  const file = square.charCodeAt(0) - 'a'.charCodeAt(0);
  const rank = Number(square[1]);
  const row = rank - 1;
  return row * NUM_SQUARES_PER_SIDE + file;
}

/** Board is 8x8 — named to avoid a magic `8` inside squareIndex/mirrorSquare. */
const NUM_SQUARES_PER_SIDE = 8;

/**
 * Mirrors a square vertically (rank r -> rank 9-r), keeping the file unchanged.
 * Used to translate a real-board move into the mover's-POV frame the model expects
 * when Black is to move (CONTRACT §a: "if Black to move, mirror the FEN first").
 */
function mirrorSquare(square: string): string {
  const file = square[0];
  const rank = NUM_SQUARES_PER_SIDE + 1 - Number(square[1]);
  return `${file}${rank}`;
}

/**
 * Mirrors a FEN piece-placement field: flips ranks top-to-bottom and swaps piece
 * colors (uppercase<->lowercase), so the side to move is always presented as "White"
 * moving up the board (CONTRACT §a, confirmed via the CSSLab reference client's
 * `preprocessMaia3` behavior — architecture only, no source copied).
 */
function mirrorPiecePlacement(piecePlacement: string): string {
  const ranks = piecePlacement.split('/');
  return [...ranks]
    .reverse()
    .map((row) => row.replace(/[a-zA-Z]/g, (c) => (c === c.toUpperCase() ? c.toLowerCase() : c.toUpperCase())))
    .join('/');
}

// ─── Board -> tensor encoding (CONTRACT §a) ─────────────────────────────────────

/**
 * Encodes a single piece-placement field into the flat (64 * 12) token tensor,
 * square-major: `tokens[squareIndex(sq) * 12 + planeIdx]`.
 */
function encodePiecePlacement(piecePlacement: string): Float32Array {
  const tokens = new Float32Array(NUM_SQUARES * PLANES_PER_SQUARE);
  const rows = piecePlacement.split('/'); // rows[0] = rank8 ... rows[7] = rank1
  for (let rowFromTop = 0; rowFromTop < NUM_SQUARES_PER_SIDE; rowFromTop++) {
    const row = NUM_SQUARES_PER_SIDE - 1 - rowFromTop; // rank8 -> row7, rank1 -> row0
    let file = 0;
    const rowStr = rows[rowFromTop] ?? '';
    for (const char of rowStr) {
      const emptyCount = Number.parseInt(char, 10);
      if (Number.isNaN(emptyCount)) {
        const planeIdx = PIECE_PLANE_ORDER.indexOf(char as (typeof PIECE_PLANE_ORDER)[number]);
        if (planeIdx >= 0) {
          tokens[(row * NUM_SQUARES_PER_SIDE + file) * PLANES_PER_SQUARE + planeIdx] = 1.0;
        }
        file += 1;
      } else {
        file += emptyCount;
      }
    }
  }
  return tokens;
}

/**
 * Encodes a FEN into the Maia-3 `tokens[64,12]` input tensor (flat, no batch dim —
 * caller/worker stacks per-ELO copies into the batch). Mirrors the board to the
 * mover's POV when Black is to move (CONTRACT §a). `historyFens` is accepted for
 * API-forward-compatibility but ignored: the confirmed contract is n=0 history
 * ("simplified" export carries no history planes — CONTRACT §a).
 */
export function encodeBoard(fen: string, historyFens?: string[]): Float32Array {
  // Accepted for API-forward-compatibility with a future non-simplified export;
  // intentionally unused under the confirmed n=0 contract (see doc comment above).
  void historyFens;
  const [piecePlacement, activeColor] = fen.split(' ');
  if (piecePlacement === undefined) {
    throw new Error(`maiaEncoding: invalid FEN (no piece-placement field): ${fen}`);
  }
  const isBlackToMove = activeColor === 'b';
  const framed = isBlackToMove ? mirrorPiecePlacement(piecePlacement) : piecePlacement;
  return encodePiecePlacement(framed);
}

// ─── ELO input (CONTRACT §b) ────────────────────────────────────────────────────

/**
 * Maps a ladder ELO value to the CONTRACT's confirmed ELO input form: a raw
 * continuous float scalar fed directly as `elo_self`/`elo_oppo` (CONTRACT §b —
 * Assumption A1 CONFIRMED, no caller-side embedding). Kept as a named function
 * (rather than passing `elo` straight through at call sites) so the single
 * confirmed mechanism has one place to change if the model's ELO input contract
 * is ever revised.
 */
export function eloToInput(elo: number): number {
  return elo;
}

// ─── Policy vocabulary index (CONTRACT §d) ──────────────────────────────────────

/**
 * Computes the flat policy-vocab index for a move, keyed by `from + to + promotion`
 * (CONTRACT §d). Queen promotions (the default) share the base from*64+to lane with
 * non-promoting moves — a pawn reaching the back rank always promotes, so `to` alone
 * disambiguates it. Underpromotions (r/b/n) use a reserved lane keyed by destination
 * + piece only (origin is unambiguous from board state). See UNDERPROMOTION_PIECE_LANES
 * doc comment for the vocab-size derivation and the open cross-validation risk.
 */
function moveVocabIndex(from: string, to: string, promotion?: string): number {
  const fromIdx = squareIndex(from);
  const toIdx = squareIndex(to);
  if (promotion === undefined || promotion === 'q') {
    return fromIdx * NUM_SQUARES + toIdx;
  }
  const laneIdx = UNDERPROMOTION_PIECE_LANES.indexOf(promotion as PromotionPiece);
  return BASE_VOCAB_SIZE + toIdx * UNDERPROMOTION_PIECE_LANES.length + laneIdx;
}

// ─── Legal-move masking + softmax (MAIA-03) ─────────────────────────────────────

/**
 * Masks the model's flat policy logits to the current FEN's legal moves (via
 * chess.js) and applies a numerically-stable softmax, returning a normalized
 * per-legal-move probability distribution keyed by SAN. Illegal moves are never
 * present in the output (only chess.js-enumerated legal moves are read from the
 * policy tensor). Mirrors from/to squares into the model's frame when Black is to
 * move (CONTRACT §d "mirror caveat") before indexing into `policy`.
 */
export function maskAndSoftmax(policy: Float32Array, fen: string): Record<string, number> {
  const chess = new Chess(fen);
  const isBlackToMove = fen.split(' ')[1] === 'b';
  const legalMoves = chess.moves({ verbose: true });

  const scores = legalMoves.map((move) => {
    const from = isBlackToMove ? mirrorSquare(move.from) : move.from;
    const to = isBlackToMove ? mirrorSquare(move.to) : move.to;
    const idx = moveVocabIndex(from, to, move.promotion);
    return policy[idx] ?? Number.NEGATIVE_INFINITY;
  });

  const max = scores.length > 0 ? Math.max(...scores) : 0;
  const exps = scores.map((s) => Math.exp(s - max));
  const sum = exps.reduce((a, b) => a + b, 0);

  const probabilities: Record<string, number> = {};
  legalMoves.forEach((move, i) => {
    probabilities[move.san] = sum > 0 ? (exps[i] ?? 0) / sum : 0;
  });
  return probabilities;
}

// ─── Expected score (CONTRACT §e, D-04) ─────────────────────────────────────────

/** Weight applied to a draw when collapsing WDL to a single expected-score fill (D-04). */
const DRAW_WEIGHT = 0.5;

/**
 * Collapses a WDL probability vector into a single expected-score fraction
 * (0..1), used by the Maia eval bar (D-04: single expected-score fill, not a
 * 3-segment W/D/L stack).
 */
export function expectedScore(wdl: WdlVector): number {
  return wdl.win + DRAW_WEIGHT * wdl.draw;
}

/**
 * Softmaxes the raw `logits_value` output into a normalized WDL probability
 * vector. Confirmed logit order (CONTRACT §e): index 0 = Loss, 1 = Draw, 2 = Win
 * — NOT W/D/L. Numerically stable (subtracts the max before exponentiating), same
 * technique as `maskAndSoftmax`. Used by `useMaiaEngine` (Task 3) to turn the
 * worker's raw per-ELO WDL logits into the vector `expectedScore` consumes.
 */
export function softmaxWdl(logits: ArrayLike<number>): WdlVector {
  const values: number[] = [];
  for (let i = 0; i < logits.length; i++) {
    values.push(logits[i] ?? 0);
  }
  const max = values.length > 0 ? Math.max(...values) : 0;
  const exps = values.map((v) => Math.exp(v - max));
  const sum = exps.reduce((a, b) => a + b, 0);
  const probabilityAt = (idx: number): number => {
    const e = exps[idx] ?? 0;
    return sum > 0 ? e / sum : 0;
  };
  return { loss: probabilityAt(0), draw: probabilityAt(1), win: probabilityAt(2) };
}

File: app/services/maia_encoding.py
"""maia_encoding.py — pure Python port of the client's Maia-3 board->tensor encoding.

Port target: frontend/src/lib/maiaEncoding.ts (and its worker mirror
frontend/public/maia/maia-worker.js). This module reproduces the CONFIRMED tensor
contract (151-MAIA-CONTRACT.md), re-verified by the 2026-07-10 headless repro that
matched the live UI root distribution to 0.01%
(.planning/notes/2026-07-10-flawchess-engine-self-execution-analysis.md):

  - 12 one-hot piece-occupancy planes per square, order white P,N,B,R,Q,K (0-5),
    black p,n,b,r,q,k (6-11); no history planes (n=0, "simplified" export).
  - square index s = row*8 + file, row = rank-1 (a1 = 0, h8 = 63).
  - When Black is to move, the piece placement is mirrored (ranks flipped + case
    swapped) BEFORE encoding, so the side to move is always presented as "White".
  - Flat, square-major token layout: tokens[squareIndex(sq)*12 + planeIdx] = 1.0.
  - Policy vocab size 4352 = 4096 base (from,to) pairs + 256 underpromotion lanes
    (64 destination squares x 4 piece lanes). Queen promotions collapse into the
    base lane. Legal-move-masked softmax over the vocab, keyed here by UCI (the
    backend's move currency) rather than the client's SAN — the probability VALUE
    is identical, only the dict key differs.
  - ELO ladder [600, 2600] step 100 (21 rungs); clamp_to_ladder_bounds mirrors the
    frontend clampToLadderBounds so a gem shown live and a gem stored in the DB
    never disagree for the same position/rating (D-04).

This module is PURE (stdlib + python-chess only, NO numpy) and imports cleanly
WITHOUT onnxruntime — so the encoding unit tests run in the default (no-group)
backend suite. numpy lives only in the isolated `maia-inference` uv group alongside
onnxruntime (GEMS-06); the ONNX session, the numpy tensor plumbing, and the parity
gate all live in scripts/maia_parity_spike.py / tests/services/test_maia_parity.py,
which sync that group. encode_board returns a plain list[float]; the ONNX callers
wrap it in a numpy float32 tensor at feed time.

Attribution: the algorithm mirrors FlawChess's own MIT maiaEncoding.ts, which was
written from scratch against the confirmed ONNX I/O contract — no CSSLab AGPL
encoding source is copied.
"""

import math
from collections.abc import Sequence

import chess

# ─── Board encoding constants (CONTRACT §a) ────────────────────────────────────

NUM_SQUARES: int = 64
PLANES_PER_SQUARE: int = 12
NUM_SQUARES_PER_SIDE: int = 8

# 12-plane order: white P,N,B,R,Q,K (0-5), black p,n,b,r,q,k (6-11).
PIECE_PLANE_ORDER: tuple[str, ...] = ("P", "N", "B", "R", "Q", "K", "p", "n", "b", "r", "q", "k")

# ─── Policy vocabulary (CONTRACT §d) ───────────────────────────────────────────

POLICY_VOCAB_SIZE: int = 4352
# Base (non-underpromotion) move-index space: every (from, to) square pair.
BASE_VOCAB_SIZE: int = NUM_SQUARES * NUM_SQUARES  # 4096
# Underpromotion piece lanes reserved per destination square (q shares the base
# lane; only r/b/n index the reserved 256 slots). 4352 - 4096 = 256 = 64 x 4.
UNDERPROMOTION_PIECE_LANES: tuple[str, ...] = ("q", "r", "b", "n")

# ─── ELO ladder (CONTRACT §c; D-04) ────────────────────────────────────────────

MAIA_ELO_LADDER_MIN: int = 600
MAIA_ELO_LADDER_MAX: int = 2600
MAIA_ELO_LADDER_STEP: int = 100


# ─── Square indexing (CONTRACT §a) ─────────────────────────────────────────────


def square_index(square: str) -> int:
    """Map an algebraic square (e.g. "e4") to its token index: s = row*8 + file,
    row = rank - 1 (a1 = 0, h8 = 63)."""
    file = ord(square[0]) - ord("a")
    rank = int(square[1])
    row = rank - 1
    return row * NUM_SQUARES_PER_SIDE + file


def mirror_square(square: str) -> str:
    """Mirror a square vertically (rank r -> rank 9-r), keeping the file unchanged.
    Used to translate a real-board move into the mover's-POV frame the model expects
    when Black is to move."""
    file = square[0]
    rank = NUM_SQUARES_PER_SIDE + 1 - int(square[1])
    return f"{file}{rank}"


def mirror_piece_placement(piece_placement: str) -> str:
    """Mirror a FEN piece-placement field: flip ranks top-to-bottom and swap piece
    colors (uppercase<->lowercase), so the side to move is always presented as
    "White" moving up the board (CONTRACT §a)."""
    ranks = piece_placement.split("/")
    return "/".join(row.swapcase() for row in reversed(ranks))


# ─── Board -> tensor encoding (CONTRACT §a) ────────────────────────────────────


def _encode_piece_placement(piece_placement: str) -> list[float]:
    """Encode a piece-placement field into a flat (64 * 12) token list, square-major:
    tokens[squareIndex(sq)*12 + planeIdx]. Plain list[float] so the module stays
    numpy-free; ONNX callers wrap it in a float32 tensor at feed time."""
    tokens = [0.0] * (NUM_SQUARES * PLANES_PER_SQUARE)
    rows = piece_placement.split("/")  # rows[0] = rank8 ... rows[7] = rank1
    for row_from_top in range(NUM_SQUARES_PER_SIDE):
        row = NUM_SQUARES_PER_SIDE - 1 - row_from_top  # rank8 -> row7, rank1 -> row0
        file = 0
        for char in rows[row_from_top]:
            if char.isdigit():
                file += int(char)
                continue
            if char in PIECE_PLANE_ORDER:
                plane_idx = PIECE_PLANE_ORDER.index(char)
                tokens[(row * NUM_SQUARES_PER_SIDE + file) * PLANES_PER_SQUARE + plane_idx] = 1.0
            file += 1
    return tokens


def encode_board(fen: str) -> list[float]:
    """Encode a FEN into the Maia-3 flat (64*12) token list. Mirrors the board to the
    mover's POV when Black is to move (CONTRACT §a). No history planes (n=0 simplified
    export)."""
    parts = fen.split(" ")
    piece_placement = parts[0]
    if not piece_placement:
        raise ValueError(f"maia_encoding: invalid FEN (no piece-placement field): {fen}")
    is_black_to_move = len(parts) > 1 and parts[1] == "b"
    framed = mirror_piece_placement(piece_placement) if is_black_to_move else piece_placement
    return _encode_piece_placement(framed)


# ─── ELO input (CONTRACT §b) ───────────────────────────────────────────────────


def elo_to_input(elo: float) -> float:
    """Map a ladder ELO value to the confirmed ELO input form: a raw continuous
    float scalar fed directly as elo_self/elo_oppo (CONTRACT §b — no embedding).
    Kept as a named function so the single confirmed mechanism has one place to
    change if the model's ELO input contract is ever revised."""
    return float(elo)


def clamp_to_ladder_bounds(rating: float) -> float:
    """Clamp a rating to the [600, 2600] ELO ladder, mirroring the frontend
    clampToLadderBounds exactly (D-04): min(MAX, max(MIN, rating))."""
    return float(min(MAIA_ELO_LADDER_MAX, max(MAIA_ELO_LADDER_MIN, rating)))


# ─── Policy vocabulary index (CONTRACT §d) ─────────────────────────────────────


def move_vocab_index(from_square: str, to_square: str, promotion: str | None) -> int:
    """Flat policy-vocab index for a move, keyed by from + to + promotion.
    Queen promotions (and non-promoting moves) share the base from*64+to lane —
    a pawn reaching the back rank always promotes, so `to` alone disambiguates it.
    Underpromotions (r/b/n) use a reserved lane keyed by destination + piece."""
    from_idx = square_index(from_square)
    to_idx = square_index(to_square)
    if promotion is None or promotion == "q":
        return from_idx * NUM_SQUARES + to_idx
    lane_idx = UNDERPROMOTION_PIECE_LANES.index(promotion)
    return BASE_VOCAB_SIZE + to_idx * len(UNDERPROMOTION_PIECE_LANES) + lane_idx


# ─── Legal-move masking + softmax (MAIA-03) ────────────────────────────────────


def mask_and_softmax(policy: Sequence[float], fen: str) -> dict[str, float]:
    """Mask the model's flat policy logits to the current FEN's legal moves (via
    python-chess) and apply a numerically-stable softmax, returning a normalized
    per-legal-move probability distribution keyed by UCI. Illegal moves are never
    present in the output. Mirrors from/to squares into the model's frame when
    Black is to move (CONTRACT §d "mirror caveat") before indexing into `policy`.

    `policy` is any indexable sequence of floats (the ONNX callers pass the model's
    logits_move output via `.tolist()`), keeping this function numpy-free."""
    board = chess.Board(fen)
    is_black_to_move = not board.turn  # chess.WHITE is True; Black to move -> False

    n = len(policy)
    ucis: list[str] = []
    scores: list[float] = []
    for move in board.legal_moves:
        from_sq = chess.square_name(move.from_square)
        to_sq = chess.square_name(move.to_square)
        if is_black_to_move:
            from_sq = mirror_square(from_sq)
            to_sq = mirror_square(to_sq)
        promotion = chess.piece_symbol(move.promotion) if move.promotion is not None else None
        idx = move_vocab_index(from_sq, to_sq, promotion)
        score = float(policy[idx]) if 0 <= idx < n else float("-inf")
        ucis.append(move.uci())
        scores.append(score)

    if not scores:
        return {}

    max_score = max(scores)
    exps = [math.exp(s - max_score) for s in scores]
    total = sum(exps)
    if total <= 0.0:
        return {uci: 0.0 for uci in ucis}
    return {uci: exp / total for uci, exp in zip(ucis, exps)}
